File: day11/SquidGrid.cs
using System.Collections.Generic;

namespace Aoc2021
{
    public class SquidGrid
    {
        public const int Size = 10;

        private readonly int[] _squids;

        private int _flashCount;
        public int FlashCount => _flashCount;

        public SquidGrid(int[] input)
        {
            _squids = new int[input.Length];
            input.CopyTo(_squids, 0);
        }

        public int ModelStep()
        {
            var flashIndices = new List<int>();
            for (int i = 0; i < _squids.Length; ++i)
            {
                if (++_squids[i] > 9)
                    flashIndices.Add(i);
            }

            int flashCount = 0;
            while (flashIndices.Count > 0)
            {
                ++flashCount;
                int index = flashIndices[0];
                flashIndices.RemoveAt(0);
                _squids[index] = 0;
                flashIndices.AddRange(IncreaseAdjacentSquids(index));
            }

            _flashCount += flashCount;
            return flashCount;
        }

        private bool IncreaseEnergy(int index)
        {
            if (_squids[index] == 0)
                return false;

            return (++_squids[index] == 10);
        }

        private IEnumerable<int> IncreaseAdjacentSquids(int index)
        {
            var flashIndices = new List<int>();
            bool hasLeft = (index % Size > 0);
            bool hasRight = (index % Size < Size - 1);

            if (hasLeft && IncreaseEnergy(index - 1))
                flashIndices.Add(index - 1);

            if (hasRight && IncreaseEnergy(index + 1))
                flashIndices.Add(index + 1);

            int offset = index - Size;
            if (offset >= 0)
            {
                if (IncreaseEnergy(offset))
                    flashIndices.Add(offset);

                if (hasLeft && IncreaseEnergy(offset - 1))
                    flashIndices.Add(offset - 1);

                if (hasRight && IncreaseEnergy(offset + 1))
                    flashIndices.Add(offset + 1);
            }

            offset = index + Size;
            if (offset < _squids.Length)
            {
                if (IncreaseEnergy(offset))
                    flashIndices.Add(offset);

                if (hasLeft && IncreaseEnergy(offset - 1))
                    flashIndices.Add(offset - 1);

                if (hasRight && IncreaseEnergy(offset + 1))
                    flashIndices.Add(offset + 1);
            }

            return flashIndices;
        }
    }
}

File: day11/squid-grid.js
export const SIZE = 10;

export function createSquidGrid(input) {
  const squids = [...input];
  let flashCount = 0;

  function increaseEnergy(index) {
    if (squids[index] == 0) {
      return false;
    }
    return (++squids[index] == 10);
  }

  function increaseAdjacentSquids(index) {
    const flashIndices = [];
    const hasLeft = (index % SIZE > 0);
    const hasRight = (index % SIZE < SIZE - 1);

    if (hasLeft && increaseEnergy(index - 1)) {
      flashIndices.push(index - 1);
    }
    if (hasRight && increaseEnergy(index + 1)) {
      flashIndices.push(index + 1);
    }

    let offset = index - SIZE;
    if (offset >= 0) {
      if (increaseEnergy(offset)) {
        flashIndices.push(offset);
      }
      if (hasLeft && increaseEnergy(offset - 1)) {
        flashIndices.push(offset - 1);
      }
      if (hasRight && increaseEnergy(offset + 1)) {
        flashIndices.push(offset + 1);
      }
    }

    offset = index + SIZE;
    if (offset < squids.length) {
      if (increaseEnergy(offset)) {
        flashIndices.push(offset);
      }
      if (hasLeft && increaseEnergy(offset - 1)) {
        flashIndices.push(offset - 1);
      }
      if (hasRight && increaseEnergy(offset + 1)) {
        flashIndices.push(offset + 1);
      }
    }

    return flashIndices;
  }

  function modelStep() {
    const flashIndices = [];
    for (let i = 0; i < squids.length; ++i) {
      if (++squids[i] > 9) {
        flashIndices.push(i);
      }
    }

    let stepFlashCount = 0;
    while (flashIndices.length) {
      ++stepFlashCount;
      const index = flashIndices.shift();
      squids[index] = 0;
      flashIndices.push(...increaseAdjacentSquids(index));
    }

    flashCount += stepFlashCount;
    return stepFlashCount;
  }

  return {
    modelStep,
    get flashCount() {
      return flashCount;
    }
  };
}
